src/pymor/algorithms/rand_la.py

```diff
@@ -2,21 +2,266 @@
+
+        .. math::
+            c_{\mathrm{est}}
+            =\frac{1}{\sqrt{2\lambda_{\mathrm{min}}}\operatorname{erf}^{-1}
+            \left(\left(\frac{\texttt{p_fail}}{n}\right)^{1/\texttt{num_testvecs}}\right)},
+
+        where :math:`\Omega` is a set of `num_testvecs` random vectors, :math:`S` denotes the inner
+        product of the range, :math:`T` denotes inner product of the source of :math:`A` and
+        :math:`n=\min\{\dim(S),\,\dim(T)\}`.
+
+        Parameters
+        ----------
+        basis_size
+            The size of the basis.
+        num_testvecs
+            Number of test vectors for error estimation.
+        p_fail
+            Maximum failure probabilty of the error estimate.
+
+        Returns
+        -------
+        err
+            The approximate error of the basis.
+        """
+        assert isinstance(basis_size, int) and basis_size > 0
+        if basis_size > min(self.A.source.dim, self.A.range.dim):
+            self.logger.warning('Requested basis is larger than the rank of the operator!')
+            self.logger.info('Proceeding with maximum operator rank.')
+            basis_size = min(self.A.source.dim, self.A.range.dim)
+        assert 0 < num_testvecs and isinstance(num_testvecs, int)
+        assert 0 < p_fail < 1
+
+        err = self._estimate_error(basis_size, num_testvecs, p_fail)
+        self.logger.info(f'Estimated error (basis dimension {basis_size}): {err:.5e}.')
+        return err
+
+    def _extend_basis(self, k):
+        W = self.A.source.random(k, distribution='normal')
+        if self.complex:
+            W += 1j * self.A.source.random(k, distribution='normal')
+
+        offset = len(self._Q[0])
+        self._Q[0].append(self.A.apply(W))
+        gram_schmidt(self._Q[0], self.range_product, offset=offset, copy=False)
+
+        for i in range(self.subspace_iterations):
+            i = 2*i + 1
+
+            k = len(self._Q[i-1]) - offset  # check if GS removed vectors
+            offset = len(self._Q[i])
+            self._Q[i].append(self._adjoint_op.apply(self._Q[i-1][-k:]))
+            gram_schmidt(self._Q[i], self.source_product, offset=offset, copy=False)
+
+            k = len(self._Q[i]) - offset  # check if GS removed vectors
+            offset = len(self._Q[i+1])
+            self._Q[i+1].append(self.A.apply(self._Q[i][-k:]))
+            gram_schmidt(self._Q[i+1], self.range_product, offset=offset, copy=False)
+
+    def _find_range(self, basis_size):
+        if basis_size > len(self._Q[-1]):
+            k = basis_size - len(self._Q[-1])
+            with self.logger.block(f'Appending {k} basis vector{"s" if k > 1 else ""} ...'):
+                self._extend_basis(k)
+            while basis_size > len(self._Q[-1]):
+                k = basis_size - len(self._Q[-1])
+                with self.logger.block(f'Appending {k} basis vector{"s" if k > 1 else ""}'
+                                       + 'to compensate for removal in gram_schmidt ...'):
```

Comment:
I think we should further think about when this can actually happen and what this says about the numerical situation. Does it really make sense to continue if a new random source vector numerically lies in the image of the preceding vectors? My intuition would be that one should actually terminate at this point.

Reply:
You are right. I did some tests with low-rank matrices and the old behaviour got caught in an infinite loop.